Fix s3 uri format so that its explicitly a s3 URI

twunderlich-grapl opened the pull request:
<!-- Thank you for your contribution to Grapl! Please take some time
to fill out this short template to help us review and understand your
PR. -->

### Which issue does this PR correspond to?
Fixes https://github.com/grapl-security/issue-tracker/issues/1045
<!-- Please provide a link to the GitHub Issue this pull request is
meant to address. -->

### What changes does this PR make to Grapl? Why?
This updates the s3 uri used with the Nomad artifact stanza in order to fix the 403 error we were seeing in AWS. This appears to be an issue with go-getter and http://, and can be seen in this thread: https://github.com/hashicorp/nomad/issues/1113
An alternate fix would be to remove the prefix entirely (ie get rid of s3:: or the previous http://)

I've removed the comments about the slack thread, since that thread has been deleted (past the 90 day threshold or whatever it is)
<!-- Please describe the functional impact of these changes and the
rationale behind them. This will help us understand your approach to
solving the problem. -->

### How were these changes tested?
1. On a nomad agent instance I downloaded the internally used go-getter binary and tested the former url and the new url:
```
sh-4.2$ ./go-getter http://plugin-registry-bucket-b1abfd2.s3.amazonaws.com/plugins/tenant_id_a7b7e3a8-fdcf-4032-a0e8-d6e3aa1973d4/plugin_type-generator/3ee133b7-8147-41bb-8908-283e0d946197.bin .
2022/10/17 23:27:21 Error downloading: bad response code: 403
sh-4.2$ ./go-getter s3::http://plugin-registry-bucket-b1abfd2.s3.amazonaws.com/plugins/tenant_id_a7b7e3a8-fdcf-4032-a0e8-d6e3aa1973d4/plugin_type-generator/3ee133b7-8147-41bb-8908-283e0d946197.bin .
2022/10/17 23:27:48 Error downloading: URL is not a valid S3 URL
sh-4.2$ ./go-getter s3::plugin-registry-bucket-b1abfd2.s3.amazonaws.com/plugins/tenant_id_a7b7e3a8-fdcf-4032-a0e8-d6e3aa1973d4/plugin_type-generator/3ee133b7-8147-41bb-8908-283e0d946197.bin .2022/10/17 23:28:52 success!
```
2. I've manually updated the Nomad job with the new uri and seen that the plugin tasks starts running successfully
<!-- Please describe how these changes were tested. There should be
sufficient detail that reviewers can replicate your testing
procedure. If the procedure is a manual one that's OK, your
description will help us work with you to get it automated. -->

Commit:
Fix s3 url format so that its explicitly s3

The hashicorp go-getter lib that Nomad uses internally doesn't
recognize the http:// uri as a s3 uri for some reason. As such, I'm
updating the format to be explicitly a s3 uri. This fixes the plugin
task when run in AWS
I've removed the comments about the slack thread, since that thread
has been deleted (past the 90 day threshold or whatever it is)

christophermaier commented on src/rust/plugin-registry/src/server/s3_url.rs:
```diff
@@ -1,13 +1,11 @@
 pub fn get_s3_url(bucket: &str, key: &str) -> String {
     let local_endpoint = std::env::var(ENV_ENDPOINT).ok();
     // If the above is specified, we're running locally against Localhost
     // Otherwise, it's against prod S3; so we can use virtual-hosted-style URLs
```
This comment looks like it needs updating now.

twunderlich-grapl replied:
Updated. Unfortunately, since I can't access the slack message, the comment I've added may be incomplete but 🤷 .
Let me know if there's something you think I should add.

christophermaier replied:
I was mainly referring to the fact that it's not an S3 virtual-host-style URL anymore (and not even a URL, or even a URI, I think :shrug:)

If it's all about how `go-getter` handles things, then a mention of that and any relevant links would be useful.

Without that kind of background information, I could _easily_ see someone in the future "helpfully" refactoring this into something that doesn't work anymore.

twunderlich-grapl replied:
Updated docs again. This should be better

christophermaier commented on src/rust/plugin-registry/src/server/s3_url.rs:
```diff
@@ -1,13 +1,11 @@
     // Otherwise, it's against prod S3; so we can use virtual-hosted-style URLs
     local_endpoint.map_or_else(
-        || format!("http://{bucket}.s3.amazonaws.com/{key}"),
+        || format!("s3::{bucket}.s3.amazonaws.com/{key}"),
```
How is this a URL? I would buy `s3://...` but this just looks _weird_.

twunderlich-grapl replied:
Ah let me change the function name to URI. And agreed that this is a very weird schema. It feels like someone was like, well our arns already use `::` as separators, why not just reuse that?

christophermaier replied:
Out of curiosity, would it have worked with `https://` instead of `http://`?

twunderlich-grapl replied:
Sadly no. I tried that early on 😿

twunderlich-grapl replied:
So apparently `s3://` does work and is officially supported by AWS. I've switched to that, especially since this is the format the s3 URI link in the console is (https://s3.console.aws.amazon.com/s3/object/plugin-registry-bucket-b1abfd2?region=us-east-1&prefix=plugins/tenant_id_024885b9-dcf0-469d-b12a-81876cc045c8/plugin_type-analyzer/51f8b573-6b7c-411c-9379-2b1070205c1f.bin and search s3 URI).
Apparently, the `s3::` was something the go-getter lib created and call [Forced Protocols](https://github.com/hashicorp/go-getter#forced-protocol). Basically its a way of saying that this is a s3 url.

Commit:
Use URI instead of URL to clarify usage

Added a comment explaining what this function is and what its used by
to replace the comment that linked to a deleted slack comversation.
I've also renamed everything from url to uri to be more explicit
about what to expect.

christophermaier commented on src/rust/plugin-registry/src/server/s3_uri.rs:
```diff
@@ -1,6 +1,9 @@
-pub fn get_s3_url(bucket: &str, key: &str) -> String {
+/// This gets a s3 URI to be used by Nomad's artifact stanza (in particular the grapl-plugin).
+/// The underlying go-getter lib, doesn't work with the `http://` schema prefix, but does work with
+/// an explicit s3 prefix `s::` or without a prefix. We've opted for the s3 prefix to be explicit.
```
`s::` -> `s3::`

Commit:
Switch to s3:// protocol and update docs

Apparently the `s3::` protocol isn't a real thing, its something the
go-getter lib created and call [Forced Protocols](https://github.com/hashicorp/go-getter#forced-protocol)
However, AWS does support a real s3 protocol `s3://`, which I've
tested with the go-getter lib and it does work. Even better, this is
the same format as the s3 URI in the console, so its easy to copy paste
a link

wimax-grapl reviewed (approved):
i think i'd prever a `grapl_config::env_helpers::is_local_aws()` but I know people have Very Strong Opinions on resurrecting `IS_LOCAL`. In this case, it makes sense to me.

christophermaier commented on src/rust/plugin-registry/src/server/s3_uri.rs:
```diff
@@ -0,0 +1,19 @@
+
+/// This gets a s3 URI to be used by Nomad's artifact stanza (in particular the grapl-plugin).
+/// The underlying library, Hashicorp's [go-getter](https://github.com/hashicorp/go-getter) lib,
+/// doesn't work with the `http://` schema prefix, but does work with an explicit s3 prefix `s://`
```
`s://` -> `s3://`

christophermaier reviewed (approved):
Tiny documentation tweak, but otherwise 👍 

Thanks for digging into the guts of this issue 🎉

Commit:
Fix typo in docs
